File: src/ae_console/js/translate.js
define('translate', [], function () {
    /* Object containing all translation strings. */
    var STRINGS = {};
    var excluded = new RegExp('^(script|style)$', 'i');
    var emptyString = new RegExp('^(\n|\s|&nbsp;)+$', 'i');

    function load (strings) {
        for (var key in strings) {
            STRINGS[key] = strings[key];
        }
    }

    /* Method to access a single translation. */
    function get (key, s1, s2) {
        try {
            if (typeof key !== 'string') { return ''; }
            // Get the string from the hash and be tolerant towards punctuation and quotes.
            var value = key;
            if (key in STRINGS) {
                value = STRINGS[key];
            } else {
                var key2 = key.replace(/[\.\:]$/, '');
                if (key2 in STRINGS) {
                    value = STRINGS[key];
                }
            }
            // Substitution of additional strings.
            for (var i = 1; i < arguments.length; i++) {
                value = value.replace('%'+(i-1), arguments[i], 'g');
            }
            return value;
        } catch (e) {
            return key || '';
        }
    }

    /* Translate the complete HTML. */
    function html (root) {
        if (root==null) { root = document.body; }
        var textNodes = [];
        var nodesWithAttr = {'title': [], 'placeholder': [], 'alt': []};
        // Translate all found text nodes.
        getTextNodes(root, textNodes, nodesWithAttr);
        for (var i = 0; i < textNodes.length; i++) {
            var text = textNodes[i].nodeValue;
            if (text.match(/^\s*$/)) { continue; }
            // Remove whitespace from the source code to make matching easier.
            var key = String(text).replace(/^(\n|\s|&nbsp;)+|(\n|\s|&nbsp;)+$/g, '');
            var value = get(key);
            // Return translated string with original whitespace.
            textNodes[i].nodeValue = text.replace(key, value);
        }
        for (var attr in nodesWithAttr) {
            for (var i = 0; i < nodesWithAttr[attr].length; i++) {
                try {
                    var node = nodesWithAttr[attr][i];
                    node.setAttribute(attr, get( node.getAttribute(attr) ) );
                } catch(e) {}
            }
        }
    }

    /* Get all text nodes that are not empty. Get also all title attributes. */
    function getTextNodes (node, textNodes, nodesWithAttr) {
        if (node && node.nodeType === 1 && !excluded.test(node.nodeName)) {
            if (node.getAttribute('title') !== null && node.getAttribute('title') !== '') { nodesWithAttr['title'].push(node); }
            if (node.getAttribute('placeholder') !== null && node.getAttribute('placeholder') !== '') { nodesWithAttr['placeholder'].push(node); }
            for (var i = 0; i < node.childNodes.length; i++) {
                var childNode = node.childNodes[i];
                if (childNode && childNode.nodeType === 3 && !emptyString.test(childNode.nodeValue)) {
                    textNodes.push(childNode);
                } else {
                    getTextNodes(childNode, textNodes, nodesWithAttr);
                }
            }
        }
    }

    return {
        get: get,
        html: html,
        load: load
    };
});

File: src/ae_console/translate.rb
module AE

  module ConsolePlugin

    class Translate

      # Load translation strings.
      # @param filename [String] a name to identify the translation file.
      def initialize(filename)
        @strings = {}
        @locale  = Sketchup.get_locale
        filepath = File.join(PATH, 'Resources', @locale, filename)
        if File.exist?(filepath)
          parse_strings(filepath)
        else
          fallback_locale = 'en-US'
          filepath        = File.join(PATH, 'Resources', fallback_locale, filename)
          if File.exist?(filepath)
            @locale = fallback_locale
            parse_strings(filepath)
          else
            puts("#{self.class}: Localization for #{@locale} not found in #{filepath}")
          end
        end
      end

      # Method to access a single translation.
      # @param key           [String]        original string in ruby script; % characters escaped by %%
      # @param substitutions [Array<String>] optional strings for substitution of %0 ... %sn
      # @return              [String]        translated string
      def get(key, *substitutions)
        key = key.to_s if key.is_a?(Symbol)
        raise(ArgumentError, 'Argument "key" must be a String or an Array of Strings.') unless key.is_a?(String) || key.nil? || (key.is_a?(Array) && key.all? { |k| k.is_a?(String) })
        return key.map { |k| self.[](k, *substitutions) } if key.is_a?(Array) # Allow batch translation of strings
        if @strings.include?(key)
          value = @strings[key].clone
        else
          Kernel.warn("warning: key #{(key.length <= 20) ? key.inspect : key[0..20].inspect + '...'} not found for locale #{@locale} (#{self.class.name})\n#{caller.first}")
          value = key.to_s.clone
        end
        # Substitution of additional strings.
        substitutions.compact.each_with_index { |substitution, i|
          value.gsub!(/\%#{i}/, substitution.to_s)
        }
        value.gsub!(/%%/, '%')
        return value.chomp
      end
      alias_method(:[], :get)

      # Push translations to a webdialog and translate all html text nodes.
      # @param [UI::WebDialog] dlg a WebDialog to translate
      #   It will translate all Text nodes and title attributes.
      #   It will also offer a JavaScript function to translate single strings.
      #   Usage: ##Translate.get(string)##
      def webdialog(dlg)
        script = %[
          requirejs(['translate'], function (Translate) {
              Translate.load(#{to_json(@strings)});
              Translate.html(document.body);
          });
        ]
        dlg.execute_script(script) unless @strings.empty?
        nil
      end

      private

      # Find translation file and parse it into a hash.
      # @param filepath [String]
      # @return [Boolean] whether the strings have been added
      def parse_strings(filepath)
        strings = {}
        File.open(filepath, 'r') { |file|
          entry      = ''
          in_comment = false
          file.each { |line|
            unless line.include?('//')
              if line.include?('/*')
                in_comment = true
              end
              if in_comment
                if line.include?('*/')
                  in_comment = false
                end
              else
                entry += line
              end
            end
            if entry.include?(';')
              keyvalue = entry.strip.gsub(/^\s*"|"\s*;$/, '').split(/"\s*=\s*"/)
              entry    = ''
              next unless keyvalue.length == 2
              key          = keyvalue[0].gsub(/\\"/, '"').gsub(/\\\\/, '\\')
              value        = keyvalue[1].gsub(/\\"/, '"').gsub(/\\\\/, '\\')
              strings[key] = value
            end
          }
        }
        @strings.merge!(strings)
        return (strings.empty?) ? false : true
      end

      def to_json(obj)
        return unless obj.is_a?(Hash)
        # remove non-JSON objects
        o = obj.reject { |k, v|
          !k.is_a?(String) && !k.is_a?(Symbol) || !v.is_a?(String) && !v.is_a?(Symbol)
        }
        # Split at every even number of unescaped quotes.
        # If it's not a string then turn Symbols into String and replace => and nil.
        json_string = o.inspect.split(/(\"(?:.*?(?:[\\][\\]+?|[^\\]))*?\")/).
        map { |s|
          if s[0..0] != '"' # If we are not inside a string
            s.gsub(/:(\S+?(?==>|\s))/, '"\\1"'). # Symbols to String
              gsub(/=>/, ':'). # Arrow to colon
              gsub(/\bnil\b/, 'null') # nil to null
          else
            s
          end
        }.join()
        return json_string
      end

    end # class Translate

  end

end
